pkg/health/health_checker.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright AppsCode Inc. and Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package health
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	dbapi "kubedb.dev/apimachinery/apis/kubedb/v1alpha2"
+
+	"k8s.io/klog/v2"
+)
+
+type HealthChecker struct {
+	healthCheckerMap map[string]healthCheckerData
+	mux              sync.Mutex
+}
+
+type healthCheckerData struct {
+	cancel            context.CancelFunc
+	ticker            *time.Ticker
+	lastPeriodSeconds int32
+}
+
+func NewHealthChecker() *HealthChecker {
+	return &HealthChecker{
+		healthCheckerMap: make(map[string]healthCheckerData),
+		mux:              sync.Mutex{},
+	}
+}
+
+// Start creates a health check go routine.
+// Call this method after successful creation of all the replicas of a database.
+func (hc *HealthChecker) Start(key string, healthCheckSpec dbapi.HealthCheckSpec, fn func(string, *HealthCard)) {
+	if healthCheckSpec.PeriodSeconds == nil || healthCheckSpec.TimeoutSeconds == nil || healthCheckSpec.FailureThreshold == nil {
```

Comment:
`*healthCheckSpec.PeriodSeconds == 0` also need to be checked, otherwise data.ticker.Reset() of line 84 will panic.

Reply:
Good catch, updated.